### PR TITLE
Reminder to document safety reasoning

### PR DESCRIPTION
**Pull request by paulocsanz**
This is the last standing issue to allow releasing a new version.

Time will be taken before release to allow new issues to be brought up and possibly continuous fuzzing to ensure it's as battle tested as the previous version, although already much safer and more ergonomic.

@Wicpar pinging you as you were very important for this release, so you can give your last thoughts before release.

### CODE REVIEW

**Comment by Wicpar** [comment]
My approach to such things is to check the safety conditions before the unsafe block to make it clear. If the code doesn't look obviously safe no amount of documentation will make sure it actually is, because we will avoid looking into the code itself in favor of reading the doc and assume it actually does what the doc says.

Also the code would be a lot clearer if we used usize internally, remove all the saturating operations.
I would also argue that it makes sense to return size and capacity as usize.

**Review by Wicpar — COMMENTED** [review]
Line 947, This is incorrect, i will fail if you replace range of max string. It must be calculated in a usize before setting the size

**Comment by paulocsanz** [comment]
Yeah, I just need to describe the safety conditions we assume, so whoever is reading the code can check that the code matches the satefy expectations described in the comment. Just using this to help me remember I need to do it before merging.

About usize `capacity()` returns usize, `len()` also returns usize.

Saturating isn't only for u8. A lot of saturating operations happen to avoid dealing with edge cases that include usize. Either subtracting from something that might be 0, or adding to an index that we haven't checked yet, so it could be usize::MAX. A few of those are unecessary but I see no benefit of using a wrapping `+/-` operator than a saturating that ensures even if we are wrong about the invariant nothing super weird will happen.

**Comment by paulocsanz** [comment]
Will take a look, thank you! This code definitely needs a thorough review and the safety comment will help reviewers!

**Comment by Wicpar** [comment]
Nevermind, overlooked you merged the usize changes, it is correct

**Comment by paulocsanz** [comment]
Line 947 here points to a `}`, will investigate further tomorrow as I have to leave for christmas diner. But I will rebase this on main as there have been a few updates.

**Comment by paulocsanz** [comment]
Oh cool

**Comment by Wicpar** [comment]
When you set the pointers fir the copy, you do not need a saturating add or sub if you know it is a valid address. Saturating operations use two instructions instead of one.

**Comment by paulocsanz** [comment]
Wrapping also uses two instructions, doesn't it? Is that the performance bottleneck of the function? I'm ok with removing that extra safety, but if it doesn't tangentially change anything in the performance I would prefer leaving it there. It's easier reasoning about the whole lib when every mathematical operation saturates.

**Comment by Wicpar** [comment]
wrapping is the default as it is just truncation of the upper bits, and is one instruction and micro-op on every architecture, except if it loads memory directly.
here is a great document that is extremely exhaustive for the x86 micro-architecture https://www.agner.org/optimize/microarchitecture.pdf

File: src/arraystring.rs [hunk 1]
```diff
@@ -873,6 +873,7 @@ where
         is_inside_boundary(self.len() + str.len() + start - end, Self::capacity())?;
 
         let this_len = self.len();
+        // TODO: document safety reasoning of this section
         unsafe {
             let ptr = self.array.as_mut_ptr();
             core::ptr::copy(
```
Comment (Wicpar) [inline src/arraystring.rs:879]: my approach here would be:
```rust
let ptr = self.array.as_mut_ptr();
let cut_start = ptr.add(start);
let cut_end = ptr.add(end);
let cut_offset = cut_start.add(str.len())
if cut_end != cut_offset {
    core::ptr::copy(cut_end, cut_offset, self.len() - end);
}
if !str.is_empty() {
    core::ptr::copy_nonoverlapping(str.as_ptr(), cut_start, str.len());
}
```

**Comment by Wicpar** [comment]
my issue with the saturating adds and subs is that it hurts readibility and induces a class of silent bugs. A lot of the room is taken up by noise, and achieves nothing in return. In pointer arithmetic it will be wrong wether it overflows or saturates... Same goes for address ranges, so size, in principle, should never have saturating add or sub. 
In debug mode, + and - will panic if it overflows which is good, we caught a bug, whereas with the saturating add/sub it will not and will likely go unnoticed.

**Comment by paulocsanz** [comment]
Solved in #18

**Pull request closed without merging**